src/server/game/map/tile/city.ts

```diff
@@ -59,4 +74,283 @@ export class City {
   removeTile(coords: Coords) {
     this.tiles.delete(coords);
   }
+
+  getUnits(): Unit[] {
+    return this.units;
+  }
+
+  getUnitPositions(): Coords[] {
+    return this.units.map(unit => unit.coords);
+  }
+
+  addUnit(unit: Unit): void {
+    this.units.push(unit);
+    if (this instanceof BarbarianCamp) {
+      unit.setBarbarian(true);
+    }
+  }
+
+  removeUnit(unit: Unit): void {
+    const unitIndex = this.units.indexOf(unit);
+    if (unitIndex > -1) {
+      this.units.splice(unitIndex, 1);
+    }
+  }
+
+  turn(world: World, map: Map): void {
+    // By default, do nothing (for now)
+  }
+
+
+}
+
+export class UnitController extends City {
+  private world: World;
+  private map: Map;
+
+  moveUnit(unit: Unit, toPos: Coords, allowCombat?: boolean): boolean {
+    const tile = this.map.getTile(toPos);
+    if (!tile) return false;
+    const movementCost = tile.getMovementCost(unit, getDirection(toPos, unit.coords));
+    if (!(unit.movement < movementCost)) {
+      if (tile.unit) {
+        if (!allowCombat) return false;
+        else {
+          this.world.meleeCombat(unit, tile.unit);
+          unit.movement = 0;
+          return true;
+        }
+      }
+      this.map.moveUnitTo(unit, toPos);
+      unit.movement -= movementCost;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  turn(world: World, map: Map): void {
+    super.turn(world, map);
+    if (!this.world) this.world = world;
+    if (!this.map) this.map = map;
+  }
+}
+
+export class BarbarianCamp extends UnitController {
+  private raidTarget?: Coords;
+  private settleTarget?: Coords;
+  private danger: boolean;
+
+  constructor(id: number, center: Coords) {
+    super(id, center, 'camp', undefined)
+    this.danger = false;
+  }
+
+  turn(world: World, map: Map): void {
+    super.turn(world, map);
+
+    const camp = map.getTile(this.center).improvement as Improvement;
+    const neighborhoodCoords = map.getNeighborsCoords(this.center, 5);
+    const neighborhoodTiles = neighborhoodCoords.map(coords => map.getTile(coords));
+    this.danger = false;
+    for (const tile of neighborhoodTiles) {
+      if (tile.unit && (tile.unit.civID !== undefined || (tile.unit.cityID !== this.id || !tile.unit.isBarbarian))) {
+        this.danger = true;
+        break;
+      }
+    }
+    if (!camp.errand) {
```

Comment:
Again, I think using early returns would make this logic easier to read. Of course you can't use it as a drop-in replacement here, but I would consider splitting each segment here into separate functions (for example, everything inside this `if` could be a function), and in that function, this could be an early return like:
```
if (camp.errand) return;
...
```